File: uhyd.py
import numpy as np
import matplotlib.pyplot as plt


class MEAS:
    def __init__(self,fname="rH_x.dat"):
        fp=open(fname,"r")
        ndat=int(fp.readline())
        print("ndat=",ndat)
        fp.readline()

        rH=[]
        hz=[]
        for row in fp:
            dat=row.strip().split(",")
            rH.append(float(dat[0]))
            hz.append(float(dat[1]))
        fp.close()

        self.rH=np.array(rH)
        self.hz=np.array(hz)
    def plot(self,ax):
        ax.plot(self.rH*100,self.hz,"ko-",markersize=8)
        ax.set_xlabel("relative humidity [%]")
        ax.set_ylabel("basal spacing [$\AA$]")
        ax.grid(True)
    def get_seg_ID(self,hz):
        if hz<self.hz[0]:
            return(-1)
        if hz>=self.hz[-1]:
            return(len(self.hz))

        indx=np.argmin(np.abs(self.hz-hz))
        h0=self.hz[indx]
        dh=hz-h0
        if dh>0:
            return(indx)
        else:
            return(indx-1)

    def rH_val(self,hz):
        i1=self.get_seg_ID(hz);
        if i1<0:
            return(self.rH[0])
        if i1>= len(self.hz):
            return(self.rH[-1])
        i2=i1+1
        xi=(hz-self.hz[i1])/(self.hz[i2]-self.hz[i1])
        eta=1.-xi
        return(self.rH[i1]*eta+self.rH[i2]*xi)

    def hz_val(self,rval):
        rH=self.rH
        hz=self.hz
        if rval<=rH[0]:
            return(hz[0])
        if rval>=rH[-1]:
            return(hz[-1])

        indx=np.argmin(np.abs(rH-rval))
        r0=self.rH[indx]
        dr=rval-r0
        if dr>0:
            i1=indx
        else:
            i1=indx-1
        i2=i1+1

        xi=(rval-rH[i1])/(rH[i2]-rH[i1])
        eta=1.-xi
        return(hz[i1]*eta+hz[i2]*xi)


class SIM:
    def __init__(self,fname):
        fp=open(fname,"r")
        fp.readline()
        hz=[]
        nH2O=[]
        dndh=[]
        for row in fp:
            data=row.strip().split(",")
            hz.append(float(data[0]))
            nH2O.append(float(data[2]))
            dndh.append(float(data[4]))

        self.hz=np.array(hz)
        self.nH2O=np.array(nH2O)
        self.dndh=np.array(dndh)

    def plot_nH2O(self,ax):
        ax.plot(self.hz,self.nH2O,"ko-",markersize=8)
        ax.set_xlabel("basal spacing [$\AA$]")
        ax.set_ylabel("n(H2O)")
        ax.grid(True)
    def plot_dndh(self,ax):
        ax.plot(self.hz,self.dndh,"ko-",markersize=8)
        ax.set_xlabel("basal spacing [$\AA$]")
        ax.set_ylabel("dn/dh[/$\AA$]")
        ax.grid(True)

    def get_seg_ID(self,hz):
        if hz<self.hz[0]:
            return(-1)
        if hz>=self.hz[-1]:
            return(len(self.hz))

        indx=np.argmin(np.abs(self.hz-hz))
        h0=self.hz[indx]
        dh=hz-h0
        if dh>0:
            return(indx)
        else:
            return(indx-1)

    def nH2O_val(self,hz):
        i1=self.get_seg_ID(hz);
        if i1<0:
            return(self.nH2O[0])
        if i1>= len(self.hz):
            return(self.nH2O[-1])
        i2=i1+1
        xi=(hz-self.hz[i1])/(self.hz[i2]-self.hz[i1])
        eta=1.-xi
        return(self.nH2O[i1]*eta+self.nH2O[i2]*xi)

    def dndh_val(self,hz):
        i1=self.get_seg_ID(hz);
        if i1<0:
            return(self.dndh[0])
        if i1>= len(self.hz):
            return(self.dndh[-1])
        i2=i1+1
        xi=(hz-self.hz[i1])/(self.hz[i2]-self.hz[i1])
        eta=1.-xi
        return(self.dndh[i1]*eta+self.dndh[i2]*xi)

    def hz_val(self,nw):
        dn=self.nH2O[1]-self.nH2O[0]
        i1=int((nw-self.nH2O[0])/dn)
        if i1<0:
            return(self.hz[0]);
        if i1>=len(self.nH2O):
            return(self.hz[-1]);

        i2=i1+1
        xi=(nw-self.nH2O[i1])/dn
        eta=1.-xi

        return(eta*self.hz[i1]+xi*self.hz[i2]);
if __name__=="__main__":
    
    plt.rcParams["font.size"]=12


    #*********** Functions of h (basal spacing) ****************
    h_xrd=MEAS(fname="rH_x.dat")    # load XRD-measured swelling curve 
    h_sim=SIM("MD_Data/mk_dndh.dat")# load MD-generated swelling curve

    hs=np.linspace(h_xrd.hz[-1],h_xrd.hz[0],201) # make h(basal spacing) axis
    dndh=[] # dn(H2O)/dh
    rH=[]   # relative humidity [0,1]
    for h in hs:    # linearly interplated data
        dndh.append(h_sim.dndh_val(h))
        rH.append(h_xrd.rH_val(h))
    dndh=np.array(dndh)
    rH=np.array(rH)
    kB=1.380649*1.e-23 #[J/K] Boltzmann constant
    NA=6.02214076*1.e23 #[/mol] Avogadro's number
    R_gas=kB*NA #[J/K/mol] Gas Constant
    T=300   # [K]
    mu_sat=-44.5*1.e03    #[J/mol]
    beta=R_gas*T/abs(mu_sat) # RT/mu_sat (R:gas constant, T:temperature, mu_sat:chemical potential at saturation)
    dUdh=(-1.0+beta*np.log(rH+1.e-08))*dndh # Gradiendt of U_hyd (hydration energy/particle)
    mu=-1+beta*np.log(rH+1.e-08)

    fig=plt.figure(figsize=[5,6])
    ax1=fig.add_subplot(311)
    ax2=fig.add_subplot(312)
    ax3=fig.add_subplot(313)
    ax1.plot(hs, dndh,"-ko",label="dn/dh")
    ax2.plot(hs,mu,"-o",label="mu/mu_sat")
    ax3.plot(hs, dUdh,"-o",label="dU/dh(h)")
    axs=[ax1,ax2,ax3]
    for ax in axs:
        ax.grid(True)
        ax.legend()
    ax2.set_ylim([-1.2,-0.9])
    ax3.set_xlabel("basal spacing $h$ [$\AA$]")


    #*********** Functions of n(H2O) (H2O molar number) ****************
    #   rH --(XRD)--> hz --(MD)--> n(H2O) 
    r1=0.0
    h=h_xrd.hz_val(r1)
    n1=h_sim.nH2O_val(h)

    r2=1.0
    h=h_xrd.hz_val(r2)
    n2=h_sim.nH2O_val(h)
    print("n(H2O) limits=",n1,n2)

    Nr=101;
    nws=np.linspace(n2,n1,Nr);   # make n(H2O) axis
    mu_n=np.zeros(Nr)
    hz_n=np.zeros(Nr)
    rh_n=np.zeros(Nr)
    k=0
    for nw in nws:
        hn=h_sim.hz_val(nw) # n(H2O) --(MD)--> hz
        rn=h_xrd.rH_val(hn) # hz --(XRD)--> rH
        mn=-1+beta*np.log(rn+1.e-08) # rH --> mu (chem. potential)
        hz_n[k]=hn
        rh_n[k]=rn
        mu_n[k]=mn
        k+=1
    mu_n+=1;        # subtract mu_sat (chem.potential at saturation)
    mu_n[-1]*=0.5   # for trapezoidal integration 
    Un=np.cumsum(mu_n)*(nws[1]-nws[0]) # integrate mu_n w.r.t. n(H2O)

    fig2=plt.figure(figsize=(5,6))
    bx1=fig2.add_subplot(311)
    bx2=fig2.add_subplot(312)
    bx3=fig2.add_subplot(313)
    bxs=[bx1,bx2,bx3]
    bx1.plot(nws,hz_n,label="h(n)")
    bx2.plot(nws,rh_n,label="R.H.(n)")
    bx3.plot(nws,Un,label="u_hyd(n)")
    for bx in bxs:
        bx.grid(True)
        bx.legend()
        bx3.set_xlabel("n(H2O)")
    bx1.set_ylabel("basal spacing [$\AA$]")
    bx2.set_ylabel("R.H.")
    bx3.set_ylabel("hydration energy")

    fig2.savefig("uhyd.png",bbox_inches="tight")
    plt.show()

    """
    nH2O=[]
    for hz in hs:
        nH2O.append(h_sim.nH2O_val(hz))
    nH2O=np.array(nH2O)

    Uh=np.cumsum(dUdh)*(hs[1]-hs[0])  #hydration energy
    fig3=plt.figure()
    bx=fig3.add_subplot(111)
    bx.plot(nH2O,Uh,"k-",linewidth=2)
    bx.grid(True)
    bx.set_xlabel("n(H2O)[mol/unit lattice]")
    bx.set_ylabel("hyndration energy/particle/$\mu_{sat}$")

    txt="# basal spacing [nm], hydration energy\n"
    ndat=len(hs)
    txt+=str(ndat)+"\n"
    for k in range(ndat):
        txt+=str(hs[ndat-1-k]*0.1)+", "+str(Uh[ndat-1-k])+"\n"

    fp=open("uhyd.dat","w")
    fp.write(txt)
    fp.close()
    plt.show()
    """

File: Uhyd.cpp
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

//	N: molar number of H2O molecules
//	H: basal spacing 

//------------------------------------------------------------
class N2H{
	public:
		int ndat;	// number of nH2O data points (equi-spaced, prescribed)
		double *nH2O;   // number of H2O/unit structure 
		double *hz;	// basal spacing [nm] (N->H map)
		int load(char fname[128]); // load data from "NaMt.dat" or "CaMt.dat"
		double dn;	// nH2O increment (for MD simlation)
		double hmin,hmax; // min and max basal spacing 
		double nmin,nmax; // min and max number of H2O mole number
		double get_h(double nw); // return h(basal spacing) for given n(H2O)
		int h_index(double h); // return index for given h(basal spacing) 
		double get_nH2O(double h);// return n(H2O) for givne h

		int ndat_h;	// equi-spaced hz data points (arbitrary) 
		double dh0;	// increment in h(basal spacing) 
		double *nH2O_intpl; // interplated n(H2O) data ( H->N map: inverse function)
		void setup_H2N(int nn); // establish H->N mapping
		void fwrite(); // export N->H(fwd.) and H->N (inv.) mappings 

	private:
	protected:
};
//------------------------------------------------------------
double N2H::get_h(double nw){

	if(nw<0.0) return(hmin);
	if(nw>=nH2O[ndat-1]) return(hmax);

	int i1,i2;
	i1=int(nw/dn);
	i2=i1+1;
	double xi=nw/dn-i1;
	double eta=1.-xi;
	return(eta*hz[i1]+xi*hz[i2]);
};
int N2H::h_index(double h){
	if(h<=hmin) return(-1);
	if(h>=hmax) return(ndat);
	int indx=0;
	while(hz[indx+1]<h) indx++;
	return(indx);
};
double N2H::get_nH2O(double h){
	int i1=N2H::h_index(h);
	if(i1<0) return(nH2O[0]);
	if(i1>=ndat) return(nH2O[ndat-1]);

	double n1=nH2O[i1];
	double n2=nH2O[i1+1];
	double xi=(h-hz[i1])/(hz[i1+1]-hz[i1]);
	double eta=1.-xi;
	return(n1*eta+n2*xi);
};
void N2H::setup_H2N(int nn){
	ndat_h=nn;
	dh0=(hmax-hmin)/(ndat_h-1);
	nH2O_intpl=(double *)malloc(sizeof(double)*ndat_h);

	int i;
	double h;
	for(i=0;i<ndat_h;i++){
		h=hmin+dh0*i;
		nH2O_intpl[i]=N2H::get_nH2O(h);
	};
};
void N2H::fwrite(){
	FILE *fp1=fopen("n2h.dat","w");
	FILE *fp2=fopen("h2n.dat","w");

	int i;
	fprintf(fp1,"%d\n",ndat);
	fprintf(fp1,"# n(H2O), h(basal spacing [nm]) \n");
	for(i=0;i<ndat;i++){
		fprintf(fp1,"%lf, %lf\n",nH2O[i],hz[i]);
	};
	double h;
	fprintf(fp2,"%d\n",ndat_h);
	fprintf(fp2,"# n(H2O), h(basal spacing [nm]) \n");
	for(i=0;i<ndat_h;i++){
		h=hmin+dh0*i;
		fprintf(fp2,"%lf, %lf\n",nH2O_intpl[i],h);
	};

	fclose(fp1);
	fclose(fp2);

};

int N2H:: load(char fname[128]){
	FILE *fp=fopen(fname,"r");
	if(fp==NULL){
		printf("File %s cannot be found !\n",fname);
		printf(" --abort process\n");
		exit(-1);
	};

	char cbff[128];
	fgets(cbff,128,fp);

	// Count data lines
	ndat=0;
	while(fgets(cbff,128,fp)!=NULL){
		ndat++;
	};
	fclose(fp);

	hz=(double *)malloc(sizeof(double)*ndat);
	nH2O=(double *)malloc(sizeof(double)*ndat);

	fp=fopen(fname,"r");
	fgets(cbff,128,fp);
	int i;
	double Un,Nc,dP;
	for(i=0;i<ndat;i++){
		fscanf(fp,"%lf, %lf, %lf, %lf, %lf\n",nH2O+i,&Un,hz+i,&Nc,&dP);
		hz[i]*=1.e-01;	// chage unit to [nm]
	};
	fclose(fp);

	dn=nH2O[1]-nH2O[0];
	hmin=hz[0];
	hmax=hz[ndat-1];
	nmin=nH2O[0];
	nmax=nH2O[ndat-1];
	return(ndat);
};
//-----------------------------------------------------------
class R2H{
	public:
	int ndat;	// number of data points
	double *rH;	// relative humidity
	double *hz;	// basal spacing [nm]
	int load(char fname[128]);
	double rH_min, rH_max;
	double hmin,hmax;
	int h_index(double h);
	int r_index(double r);
	double get_rH(double h);
	double get_hz(double r);

	private:
	protected:
};
//-----------------------------------------------------------
int R2H::r_index(double r){
	if(r<=rH_min) return(-1);
	if(r>=rH_max) return(ndat);
	int indx=0;
	while(rH[indx+1]<r) indx++;
	return(indx);
}
double R2H::get_hz(double r){	// rH -> h mapping
	int i1=R2H::r_index(r);
	if(i1<0) return(hz[0]);
	if(i1>=ndat) return(hz[ndat-1]);

	double h1=hz[i1];
	double h2=hz[i1+1];
	double xi=(r-rH[i1])/(rH[i1+1]-rH[i1]);
	double eta=1.-xi;
	return(h1*eta+h2*xi);
};

int R2H::h_index(double h){
	if(h<=hmin) return(-1);
	if(h>=hmax) return(ndat);
	int indx=0;
	while(hz[indx+1]<h) indx++;
	return(indx);
};
double R2H::get_rH(double h){	// h -> rH mapping
	int i1=R2H::h_index(h);
	if(i1<0) return(rH[0]);
	if(i1>=ndat) return(rH[ndat-1]);

	double rH1=rH[i1];
	double rH2=rH[i1+1];
	double xi=(h-hz[i1])/(hz[i1+1]-hz[i1]);
	double eta=1.-xi;
	return(rH1*eta+rH2*xi);
};
int R2H::load(char fname[128]){
	FILE *fp=fopen(fname,"r");
	if(fp==NULL){
		printf("File %s cannot be found !\n",fname);
		printf(" --abort process\n");
		exit(-1);
	};
	char cbff[128];
	fscanf(fp,"%d\n",&ndat);
	fgets(cbff,128,fp);

	hz=(double *)malloc(sizeof(double)*ndat);
	rH=(double *)malloc(sizeof(double)*ndat);
	int i;
	double tmp;
	for(i=0;i<ndat;i++){
		fscanf(fp,"%lf, %lf, %lf\n",rH+i,hz+i,&tmp);
		hz[i]*=1.e-01;
	};

	rH_min=rH[0];
	rH_max=rH[ndat-1];
	hmin=hz[0];
	hmax=hz[ndat-1];

	//printf("rH_lim=%lf %lf\n",rH_min,rH_max);
	//printf("hz_lim=%lf %lf\n",hmin,hmax);
	return(ndat);
};

//-----------------------------------------------------------
int main(){

	char fname1[128]="MD_Data/NaMt.dat";
	char fname2[128]="rH_x.dat";
	char fnout[128]="Uhyd.out";

	N2H n2h_MD;	// MD-generated swelling curve
	n2h_MD.load(fname1);

	R2H r2h_XRD;	// XRD-meassured swelling curve
	r2h_XRD.load(fname2);

	// rH --> hz --> n(H2O)
	int i,N=101;
	double drH=1./(N-1);
	double hz_r,nw_r;
	for(i=0;i<N;i++){
		hz_r=r2h_XRD.get_hz(i*drH);
		nw_r=n2h_MD.get_nH2O(hz_r);
	};

	i=0;
	hz_r=r2h_XRD.get_hz(i*drH);
	nw_r=n2h_MD.get_nH2O(hz_r);
	double nmin=nw_r;
	i=N;
	hz_r=r2h_XRD.get_hz(i*drH);
	nw_r=n2h_MD.get_nH2O(hz_r);
	double nmax=nw_r;


	double dn=(nmax-nmin)/(N-1);
	double nw; //,hz_n,rh_n,mu; 

    	double kB=1.380649*1.e-23;  // [J/K] Boltzmann constant
    	double NA=6.02214076*1.e23; // [/mol] Avogadro's number
	double R_gas=kB*NA; //[J/K/mol] Gas Constant
    	double T=300;   // [K] Tempearture
	double mu_sat=-44.5*1.e03;  // [J/mol] chem. potential at vapor saturation
    	double beta=R_gas*T/abs(mu_sat); // RT/mu_sat 
	double dGn;
	double mu_prev=0.0;
	double mu0=abs(mu_sat)*1.e-03; // [kJ/mol]

	double *hz_n=(double *)malloc(sizeof(double)*N);
	double *rh_n=(double *)malloc(sizeof(double)*N);
	double *mu_var=(double *)malloc(sizeof(double)*N);
	double *Gn=(double *)malloc(sizeof(double)*N);
	double *Gn_sat=(double *)malloc(sizeof(double)*N);
	double Gvar=0.0,Gsat=0.0;
	for(i=N-1;i>=0;i--){
		nw=nmin+dn*i;
		hz_n[i]=n2h_MD.get_h(nw); // MD-swelling curve
		rh_n[i]=r2h_XRD.get_rH(hz_n[i]);
		//mu=-1.+beta*log(rh_n+1.e-08);
		mu_var[i]=R_gas*T*log(rh_n[i]+1.e-08);	
		Gn[i]=Gvar;
		Gn_sat[i]=Gsat;
		dGn=-(mu_var[i]+mu_prev)*0.5*dn;
		Gvar+=dGn;
		Gsat+=(mu0*dn);
		mu_prev=mu_var[i];
	};
	FILE *fp=fopen(fnout,"w");
	fprintf(fp,"%d\n",N);
	fprintf(fp,"# n(H2O), hz[nm], R.H., mu_var, mu_sat(kJ/mol), del{G(n)}(nolinear part of G), G(n)_sat[kJ/mol]\n");
	for(i=0;i<N;i++){
		nw=nmin+dn*i;
		fprintf(fp,"%lf, %lf, %lf, %lf, %lf, %lf, %lf\n",nw,hz_n[i],rh_n[i],mu_var[i]*1.e-03,mu_sat*1.e-03,Gn[i]*1.e-03,Gn_sat[i]);
	}

	printf("Output File --> %s\n",fnout);

	return(0);
};
